hledger-lib/Hledger/Read/Common.hs

```diff
@@ -220,29 +227,50 @@ definputopts = InputOpts
     , new_               = False
     , new_save_          = True
     , pivot_             = ""
+    , forecast_          = Nothing
     , auto_              = False
     , balancingopts_     = def
     , strict_            = False
     }
 
-rawOptsToInputOpts :: RawOpts -> InputOpts
-rawOptsToInputOpts rawopts = InputOpts{
-     -- files_             = listofstringopt "file" rawopts
-     mformat_           = Nothing
-    ,mrules_file_       = maybestringopt "rules-file" rawopts
-    ,aliases_           = listofstringopt "alias" rawopts
-    ,anon_              = boolopt "anon" rawopts
-    ,new_               = boolopt "new" rawopts
-    ,new_save_          = True
-    ,pivot_             = stringopt "pivot" rawopts
-    ,auto_              = boolopt "auto" rawopts
-    ,balancingopts_     = def{ ignore_assertions_ = boolopt "ignore-assertions" rawopts
-                             , infer_prices_      = not noinferprice
-                             }
-    ,strict_            = boolopt "strict" rawopts
-    }
+rawOptsToInputOpts :: RawOpts -> IO InputOpts
+rawOptsToInputOpts rawopts = do
+    d <- getCurrentDay
+
+    return InputOpts{
+       -- files_             = listofstringopt "file" rawopts
+       mformat_           = Nothing
+      ,mrules_file_       = maybestringopt "rules-file" rawopts
+      ,aliases_           = listofstringopt "alias" rawopts
+      ,anon_              = boolopt "anon" rawopts
+      ,new_               = boolopt "new" rawopts
+      ,new_save_          = True
+      ,pivot_             = stringopt "pivot" rawopts
+      ,forecast_          = forecastPeriodFromRawOpts d rawopts
+      ,auto_              = boolopt "auto" rawopts
+      ,balancingopts_     = def{ ignore_assertions_ = boolopt "ignore-assertions" rawopts
+                               , infer_prices_      = not noinferprice
+                               }
+      ,strict_            = boolopt "strict" rawopts
+      }
   where noinferprice = boolopt "strict" rawopts || stringopt "args" rawopts == "balancednoautoconversion"
 
+-- | get period expression from --forecast option
+forecastPeriodFromRawOpts :: Day -> RawOpts -> Maybe DateSpan
+forecastPeriodFromRawOpts d opts = case maybestringopt "forecast" opts of
+    Nothing -> Nothing
+    Just "" -> Just $ DateSpan Nothing mspecifiedend
+    Just str ->
+      either (\e -> usageError $ "could not parse forecast period : "++customErrorBundlePretty e) (Just . snd) $ 
+      parsePeriodExpr d $ stripquotes $ T.pack str
+  where
+    -- "They end on or before the specified report end date, or 180 days from today if unspecified."
+    mspecifiedend = dbg2 "specifieddates" . fmap (addDays 1) $ reportPeriodLastDay =<< rspec
+    -- Ugly hack. We just need the requested end date; do we really need to calculate the whole
+    -- rspec from scratch? Coerce errors to Maybe because any errors should be handled elsewhere,
+    -- but this is not the correct behaviour.
```

Comment:
This all sounds rather obscure and possibly error-hiding (the last sentence). Maybe it's an old comment ? It sounds like something that needs work so let's mark it with XXX, and let's clarify the consequences if any, especially if user-visible

Reply:
Not very old, I just added it. I certainly don't want this merged as is, but am having trouble coming up with a more elegant solution.

Reply:
I've found a better solution, though it's still not ideal. No more questionable error handling required.